arch: arm: Add symbol for flash used

hakonfam opened the pull request:
Add symbol which contains the number of bytes contained
in the image.

Using '_image_rom_end' will not work, as there are
symbols loaded after its value.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

hakonfam committed:
arch: arm: Add symbol for flash used

Add symbol which contains the number of bytes contained
in the image.

Using '_image_rom_end' will not work, as there are
symbols loaded after its value.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

ioannisg commented:
@SebastianBoe could you also take a look?

ioannisg reviewed (changes requested):
Could you, simply, move _image_rom_end to the _real_ end of the linker script (ROMABLE_REGION), and make a comment that this must indicate end-of-used-code-section?
Perhaps, you might have to move the GROUP_END, too.

The less new symbols we introduced, the safer and cleaner it becomes, IMHO.

Also: could you use "arch: arm: " as prefix for the commit title?

thanks

hakonfam commented:
> Could you, simply, move _image_rom_end to the real end of the linker script (ROMABLE_REGION), and make a comment that this must indicate end-of-used-code-section?
Perhaps, you might have to move the GROUP_END, too.

I could, but that would change the semantics of this symbol. This is currently being used by `elf_helper.py`. There is no documentation about this symbol, so the intended use might be correct as it is today. Also, moving `GROUP_END` might have unknown effects (even though it looks like an empty macro).

If the zephyr chiefs still wants me to move the existing symbol instead, its no problem for me.

ioannisg commented:
> 
> 
> > Could you, simply, move _image_rom_end to the real end of the linker script (ROMABLE_REGION), and make a comment that this must indicate end-of-used-code-section?
> > Perhaps, you might have to move the GROUP_END, too.
> 
> I could, but that would change the semantics of this symbol. This is currently being used by `elf_helper.py`. There is no documentation about this symbol, so the intended use might be correct as it is today. Also, moving `GROUP_END` might have unknown effects (even though it looks like an empty macro).
> 
> If the zephyr chiefs still wants me to move the existing symbol instead, its no problem for me.
> 
> Changed commit title to `arch: arm`

It is fine for me, if you move _image_rom_end to the end of code section; this is exactly what the name suggests. I would skip adding symbols if we can avoid it.

Adding @AdithyaBaglody to help out here.

AdithyaBaglody commented:
> It is fine for me, if you move _image_rom_end to the end of code section; this is exactly what the name suggests. I would skip adding symbols if we can avoid it.

I totally agree for not adding new symbols but i think in this case we might not be able to get the required outcome.
Lets say we move the image_rom_end to the bottom. 
We cant be sure that the address we get is a Load address and not the link address.
The line counter is pretty tricky when we use the AT command.

AdithyaBaglody commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -481,4 +481,12 @@ SECTIONS
 	KEEP(*(.gnu.attributes))
 	}
 
+SECTION_PROLOGUE(.last_section,(NOLOAD),)
```
Curious here.
Can we create empty sections and ask it to give the load address when there is no data to load?

hakonfam replied:
Not sure what the specification says about this, but it works as I intended.
The empty section will still be placed in the output, with zero length.
Hence we can use its load address.

Last lines in .map file:
```
last_section 0xb3f0 0x0
             0xb3f0    _flash_used = (LOADADDR (.last_section) - _image_rom_start)
```

The

carlescufi commented:
@AdithyaBaglody @ioannisg can we come to a conclusion on this, either select another approach or approve?

AdithyaBaglody reviewed (approved):
I am not aware of a better way to do this. Although I am not a fan of empty sections. I think we can go ahead with this.

ioannisg reviewed (commented):
If we are to go this way, we need the following, IMO:
- this empty section must be last in ROMMABLE area, right? If so, we need documentation before the section Prologue, stating that this section must be last. 
- ideally, we should also document why this empty section is needed (why the _image_rom_end does not work)
- expressions -> expression
- don't we need to export the _flash_used symbol in linker-defs.h?
-

hakonfam commented:
> this empty section must be last in ROMMABLE area, right? If so, we need documentation before the section Prologue, stating that this section must be last.

Fixed

> ideally, we should also document why this empty section is needed (why the _image_rom_end does not work)

Fixed, I don't know why moving `_image_rom_end` does not work, since it was not documented in the first place, and I don't know its expected behavior. 

> expressions -> expression

Fixed.

> don't we need to export the _flash_used symbol in linker-defs.h?

No, it is available for reference as it is now.

ioannisg commented on include/arch/arm/cortex_m/scripts/linker.ld:
```diff
@@ -481,4 +481,13 @@ SECTIONS
 	KEEP(*(.gnu.attributes))
 	}
 
+/* This section must be at the end of the linker script */
```
Perhaps:     /* Must be liastin romable region */
This is similar to 

    /* Must be first in ramable region */

hakonfam commented:
> > No, it is available for reference as it is now.
> 
> So, AFAICS, all the linker symbols that are used in C code are declared in linker-defs.h.
> I don't see, clearly, why this should not be the case for your _flash_used symbol. Any idea?
> 
> Ideally, it should be there, along with some documentation; please, see how other symbols are declared and documented.

I will add it.

hakonfam commented:
Check now @ioannisg

ioannisg reviewed (commented):
Looks ok now - I asked @SebastianBoe to have a look, too.

SebastianBoe commented:
LGTM.

"_image_rom_end" and "_flash_used" appear to be pointing to two different things. One points to a VMA, and the other an LMA. There exists use-cases for referencing both.
